File: Data Structures & Algorithms/house-robber-ii/submission-4.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        # one = two = 0

        # for i in range(len(nums)):
        #     tmp =  max(nums[i] + one, two)
        #     one = two
        #     two = tmp
        if len(nums) == 1:
            return nums[0]
        return max(self.linear_rob(nums[1:]), self.linear_rob(nums[:-1]))
        
    def linear_rob(self, nums):
        one = two = 0
        for i in range(len(nums)):
            tmp = max(nums[i] + one, two)
            one = two
            two = tmp

        return two

File: Data Structures & Algorithms/climbing-stairs/submission-4.java
class Solution {
    public int climbStairs(int n) {
        int s1 = 1;
        int s2 = 2;

        if (n < 3) {
            return n;
        }

        for (int x = 2; x < n; x++) {
            int tmp = s1 + s2;
            s1 = s2;
            s2 = tmp;
        }
        return s2;
    }
}
